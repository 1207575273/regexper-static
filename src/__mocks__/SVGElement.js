import React from 'react';

import Base from 'components/SVG/Base';

class SVGElement extends Base {
  reflow() {
    return this.setBBox(this.props.bbox);
  }

  render() {
    return <text>Mock content</text>;
  }
}

export default SVGElement;
