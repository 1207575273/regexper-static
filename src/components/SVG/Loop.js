import React from 'react';
import PropTypes from 'prop-types';

import Base from './Base';
import style from './style';
import Path from './path';

const skipPath = (box, greedy) => {
  const vert = Math.max(0, box.axisY - 10);
  const horiz = box.width - 10;

  let path = new Path({ relative: true });

  if (!greedy) {
    path
      .moveTo({ x: 10, y: box.axisY + box.offsetY - 15, relative: false })
      .lineTo({ x: 5, y: 5 })
      .moveTo({ x: -5, y: -5 })
      .lineTo({ x: -5, y: 5 });
  }

  return path
    .moveTo({ x: 0, y: box.axisY + box.offsetY, relative: false })
    .quadraticCurveTo({ cx: 10, cy: 0, x: 10, y: -10 })
    .lineTo({ y: -vert })
    .quadraticCurveTo({ cx: 0, cy: -10, x: 10, y: -10 })
    .lineTo({ x: horiz })
    .quadraticCurveTo({ cx: 10, cy: 0, x: 10, y: 10 })
    .lineTo({ y: vert })
    .quadraticCurveTo({ cx: 0, cy: 10, x: 10, y: 10 });
};

const repeatPath = (box, greedy) => {
  const vert = box.height - box.axisY - 10;

  let path = new Path({ relative: true });

  if (greedy) {
    path
      .moveTo({ x: box.offsetX + box.width + 10, y: box.axisY + box.offsetY + 15, relative: false })
      .lineTo({ x: 5, y: -5 })
      .moveTo({ x: -5, y: 5 })
      .lineTo({ x: -5, y: -5 });
  }

  return path
    .moveTo({ x: box.offsetX, y: box.axisY + box.offsetY, relative: false })
    .quadraticCurveTo({ cx: -10, cy: 0, x: -10, y: 10 })
    .lineTo({ y: vert })
    .quadraticCurveTo({ cx: 0, cy: 10, x: 10, y: 10 })
    .lineTo({ x: box.width })
    .quadraticCurveTo({ cx: 10, cy: 0, x: 10, y: -10 })
    .lineTo({ y: -vert })
    .quadraticCurveTo({ cx: 0, cy: -10, x: -10, y: -10 });
};

/** @extends React.PureComponent */
class Loop extends Base {
  get contentOffset() {
    const { skip, repeat } = this.props;

    if (skip) {
      return { x: 15, y: 10 };
    } else if (repeat) {
      return { x: 10, y: 0 };
    } else {
      return { x: 0, y: 0 };
    }
  }

  preReflow() {
    return this.contained;
  }

  reflow() {
    return new Promise(resolve => {
      const { skip, repeat, greedy } = this.props;
      const box = this.contained.getBBox();
      const labelBox = this.label ? this.label.getBBox() : { width: 0, height: 0 };

      let height = box.height + labelBox.height;
      if (skip) {
        height += 10;
      }
      if (repeat) {
        height += 10;
      }

      this.setBBox({
        width: box.width + this.contentOffset.x * 2,
        height,
        axisY: box.axisY + this.contentOffset.y,
        axisX1: box.axisX1 + this.contentOffset.x,
        axisX2: box.axisX2 + this.contentOffset.x
      });

      box.offsetX = this.contentOffset.x;
      box.offsetY = this.contentOffset.y;

      this.setState({
        labelTransform: `translate(${ this.getBBox().width - labelBox.width - 10 } ${ this.getBBox().height + 2 })`,
        loopPaths: [
          skip && skipPath(box, greedy),
          repeat && repeatPath(box, greedy)
        ].filter(Boolean).join('')
      }, resolve);
    });
  }

  containedRef = contained => this.contained = contained

  labelRef = label => this.label = label

  render() {
    const { label, children } = this.props;
    const { loopPaths, labelTransform } = this.state || {};

    const textProps = {
      transform: labelTransform,
      style: style.infoText,
      ref: this.labelRef
    };

    return <React.Fragment>
      <path d={ loopPaths } style={ style.connectors }></path>
      { label && <text { ...textProps }>{ label }</text> }
      <g transform={ `translate(${ this.contentOffset.x } ${ this.contentOffset.y })` }>
        { React.cloneElement(React.Children.only(children), {
          ref: this.containedRef
        }) }
      </g>
    </React.Fragment>;
  }
}

Loop.propTypes = {
  children: PropTypes.node.isRequired,
  greedy: PropTypes.bool,
  label: PropTypes.string,
  skip: PropTypes.bool,
  repeat: PropTypes.bool
};

export default Loop;
