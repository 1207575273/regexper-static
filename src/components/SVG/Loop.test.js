import React from 'react';
import { mount } from 'enzyme';

import Loop from './Loop';

import SVGElement from '__mocks__/SVGElement';

const originalGetBBox = window.Element.prototype.getBBox;

describe('Loop', () => {
  beforeEach(() => {
    window.Element.prototype.getBBox = function() {
      return { width: 100, height: 10 };
    };
  });

  afterEach(() => {
    window.Element.prototype.getBBox = originalGetBBox;
  });

  test('rendering', async () => {
    const component = mount(
      <Loop>
        <SVGElement bbox={{ width: 100, height: 100 }}/>
      </Loop>
    );
    await component.instance().doReflow();
    component.update();
    expect(component).toMatchSnapshot();
  });

  test('rendering with skip path', async () => {
    const component = mount(
      <Loop skip>
        <SVGElement bbox={{ width: 100, height: 100 }}/>
      </Loop>
    );
    await component.instance().doReflow();
    component.update();
    expect(component).toMatchSnapshot();
  });

  test('rendering with repeat path', async () => {
    const component = mount(
      <Loop repeat>
        <SVGElement bbox={{ width: 100, height: 100 }}/>
      </Loop>
    );
    await component.instance().doReflow();
    component.update();
    expect(component).toMatchSnapshot();
  });

  test('rendering with repeat path and label', async () => {
    const component = mount(
      <Loop repeat label="Test label">
        <SVGElement bbox={{ width: 100, height: 100 }}/>
      </Loop>
    );
    await component.instance().doReflow();
    component.update();
    expect(component).toMatchSnapshot();
  });

  test('rendering with skip and repeat paths', async () => {
    const component = mount(
      <Loop skip repeat>
        <SVGElement bbox={{ width: 100, height: 100 }}/>
      </Loop>
    );
    await component.instance().doReflow();
    component.update();
    expect(component).toMatchSnapshot();
  });

  test('rendering with greedy skip and repeat paths', async () => {
    const component = mount(
      <Loop greedy skip repeat>
        <SVGElement bbox={{ width: 100, height: 100 }}/>
      </Loop>
    );
    await component.instance().doReflow();
    component.update();
    expect(component).toMatchSnapshot();
  });
});
