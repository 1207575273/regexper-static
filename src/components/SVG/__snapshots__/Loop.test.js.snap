// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loop rendering 1`] = `
<Loop>
  <path
    d=""
    style={
      Object {
        "fillOpacity": 0,
        "stroke": "#000",
        "strokeWidth": "2px",
      }
    }
  />
  <g
    transform="translate(0 0)"
  >
    <SVGElement
      bbox={
        Object {
          "height": 100,
          "width": 100,
        }
      }
    >
      <text>
        Mock content
      </text>
    </SVGElement>
  </g>
</Loop>
`;

exports[`Loop rendering with greedy skip and repeat paths 1`] = `
<Loop
  greedy={true}
  repeat={true}
  skip={true}
>
  <path
    d="M0,60q10,0 10,-10v-40q0,-10 10,-10h90q10,0 10,10v40q0,10 10,10M125,75l5,-5m-5,5l-5,-5M15,60q-10,0 -10,10v40q0,10 10,10h100q10,0 10,-10v-40q0,-10 -10,-10"
    style={
      Object {
        "fillOpacity": 0,
        "stroke": "#000",
        "strokeWidth": "2px",
      }
    }
  />
  <g
    transform="translate(15 10)"
  >
    <SVGElement
      bbox={
        Object {
          "height": 100,
          "width": 100,
        }
      }
    >
      <text>
        Mock content
      </text>
    </SVGElement>
  </g>
</Loop>
`;

exports[`Loop rendering with repeat path 1`] = `
<Loop
  repeat={true}
>
  <path
    d="M10,50q-10,0 -10,10v40q0,10 10,10h100q10,0 10,-10v-40q0,-10 -10,-10"
    style={
      Object {
        "fillOpacity": 0,
        "stroke": "#000",
        "strokeWidth": "2px",
      }
    }
  />
  <g
    transform="translate(10 0)"
  >
    <SVGElement
      bbox={
        Object {
          "height": 100,
          "width": 100,
        }
      }
    >
      <text>
        Mock content
      </text>
    </SVGElement>
  </g>
</Loop>
`;

exports[`Loop rendering with repeat path and label 1`] = `
<Loop
  label="Test label"
  repeat={true}
>
  <path
    d="M10,50q-10,0 -10,10v40q0,10 10,10h100q10,0 10,-10v-40q0,-10 -10,-10"
    style={
      Object {
        "fillOpacity": 0,
        "stroke": "#000",
        "strokeWidth": "2px",
      }
    }
  />
  <text
    style={
      Object {
        "dominantBaseline": "text-after-edge",
        "fontFamily": "Arial",
        "fontSize": "12px",
      }
    }
    transform="translate(10 122)"
  >
    Test label
  </text>
  <g
    transform="translate(10 0)"
  >
    <SVGElement
      bbox={
        Object {
          "height": 100,
          "width": 100,
        }
      }
    >
      <text>
        Mock content
      </text>
    </SVGElement>
  </g>
</Loop>
`;

exports[`Loop rendering with skip and repeat paths 1`] = `
<Loop
  repeat={true}
  skip={true}
>
  <path
    d="M10,45l5,5m-5,-5l-5,5M0,60q10,0 10,-10v-40q0,-10 10,-10h90q10,0 10,10v40q0,10 10,10M15,60q-10,0 -10,10v40q0,10 10,10h100q10,0 10,-10v-40q0,-10 -10,-10"
    style={
      Object {
        "fillOpacity": 0,
        "stroke": "#000",
        "strokeWidth": "2px",
      }
    }
  />
  <g
    transform="translate(15 10)"
  >
    <SVGElement
      bbox={
        Object {
          "height": 100,
          "width": 100,
        }
      }
    >
      <text>
        Mock content
      </text>
    </SVGElement>
  </g>
</Loop>
`;

exports[`Loop rendering with skip path 1`] = `
<Loop
  skip={true}
>
  <path
    d="M10,45l5,5m-5,-5l-5,5M0,60q10,0 10,-10v-40q0,-10 10,-10h90q10,0 10,10v40q0,10 10,10"
    style={
      Object {
        "fillOpacity": 0,
        "stroke": "#000",
        "strokeWidth": "2px",
      }
    }
  />
  <g
    transform="translate(15 10)"
  >
    <SVGElement
      bbox={
        Object {
          "height": 100,
          "width": 100,
        }
      }
    >
      <text>
        Mock content
      </text>
    </SVGElement>
  </g>
</Loop>
`;
